### ISSUE TITLE
gethostbyaddr() wrong behaviour

### ISSUE THREAD

**Issue by chemelli74** [alpinelinux/docker-alpine#107]
Hi,

I'm a [Home Assistant](https://www.home-assistant.io/) user.
I'm experiencing a issue that I tracked down to Alpine docker image.

You can find all info here: [Reverse DNS caching](https://github.com/home-assistant/core/issues/39040)

Let me know if I can be of any help debugging this behaviour.

Simone

**Comment by chemelli74** [comment]
Anyone can help looking into this ?

Thank you,

Simone

**Comment by ncopa** [comment]
I'm not sure I understand what the problem is, nor do I know how to reproduce it?

**Comment by chemelli74** [comment]
Sorry if I was too concise in my descrition.
I try to reword it a bit.

I have a issue with reverse DNS on Alpine/Docker that I don't have on Ubuntu/Docker on the same host.

Steps to reproduce:


- Create gethostbyaddr.py with the following code:

```
import socket

res = socket.gethostbyaddr('192.168.xxx.yy')
print(res)
```

- Run it in a linux container (I have a vanilla ubuntu) and record answer:

('shelly-named-by-me.fritz.box', [], ['192.168.xxx.yy'])

- Run same in Alpine container and record answer:

('shellyswitch25-XXXXXX.fritz.box', ['shellyswitch25-XXXXXX.fritz.box'], ['192.168.xxx.yy'])

As you can see the answer is different.
Take care that "shelly-named-by-me.fritz.box" is the current name of the device, while "shellyswitch25-XXXXXX.fritz.box" is the original device name that no longer exist on the network.

Seems there is some kind of caching on Alpine that stores the old name, and that survive a system restart.

Let me know if now is more clear.

Simone

**Comment by chemelli74** [comment]
Hi @ncopa , was above explanation clear enough ?

Let me know how I can help progressing with this issue.
Thx,

Simone

**Comment by chemelli74** [comment]
Hi, any news on this ?

Thank you in advance,

Simone

**Comment by chemelli74** [comment]
Hi @ncopa, hi Team,

anyone that can look at this issue and help understanding what's happening ?

Simone

**Comment by ncopa** [comment]
I don't think alpine or musl is caching anything so this depends completely on what the dns server responds.

What do you get from `dig -x 192.168.xx.yy`?

**Comment by chemelli74** [comment]
Hi, thx for the reply.
Please see the 2 tests and the different results:

gethostbyaddr query:
```
bash-5.0# python gethostbyaddr.py
('shellyswitch25-xxxxxx.fritz.box', ['shellyswitch25-xxxxxx.fritz.box'], ['192.168.yyy.xx'])
```

dig query:
```
bash-5.0# dig -x 192.168.yyy.xx

; <<>> DiG 9.16.6 <<>> -x 192.168.yyy.xx
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 38471
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 1, ADDITIONAL: 2

;; QUESTION SECTION:
;xx.yyy.168.192.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
xx.yyy.168.192.in-addr.arpa. 9  IN      PTR     shelly-camera-degli-ospiti-tapparella.fritz.box.
xx.yyy.168.192.in-addr.arpa. 9  IN      PTR     shellyswitch25-xxxxxx.fritz.box.

;; AUTHORITY SECTION:
xx.yyy.168.192.in-addr.arpa. 9  IN      NS      fritz.box.

;; ADDITIONAL SECTION:
fritz.box.              9       IN      A       192.168.yyy.zzz

;; Query time: 3 msec
;; SERVER: 192.168.yyy.zzz#53(192.168.yyy.zzz)
;; WHEN: Wed Dec 02 15:41:20 CET 2020
;; MSG SIZE  rcvd: 209
```

dig is working as expected.

Simone

**Comment by ncopa** [comment]
> ;; ANSWER SECTION:
xx.yyy.168.192.in-addr.arpa. 9  IN      PTR     shelly-camera-degli-ospiti-tapparella.fritz.box.
xx.yyy.168.192.in-addr.arpa. 9  IN      PTR     shellyswitch25-xxxxxx.fritz.box.

Looks like DNS server returns both names, so both test cases are technically correct, even if they are different.

If `shellyswitch25-XXXXXX.fritz.box` is wrong, then the DNS server is giving you a wrong response and you need configure your DNS server.

I don't see why we should change Alpine in this case. I mean I understand the inconvenience that it is different than glibc, but what is the reason that it is not glibc that should change?

What does other OSes, like openbsd, freebsd, macOS return?

**Comment by chemelli74** [comment]
>
> What does other OSes, like openbsd, freebsd, macOS return?
>

All system I tested with return both values.

Only gethostbyaddr() on Alpine return 2 times the first value and not the second one.

Simone

**Comment by ncopa** [comment]
In either case, musl just gives you one of the two resource records from the DNS server response. There is nothing wrong with what alpine does here.

**Comment by chemelli74** [comment]
> In either case, musl just gives you one of the two resource records from the DNS server response. There is nothing wrong with what alpine does here.

I'm sorry but this is not true.
If DNS has 2 records, then gethostbyaddr should report both and not only 1.
How it choice the one to report ? What about the other ? Why should be ok to ignore it ?

Simone

**Comment by ncopa** [comment]
> If DNS has 2 records, then gethostbyaddr should report both and not only 1.

Do you have any reference or any other documentation that says it should return multiple hostnames? Also ubuntu only returns a single hostname, but the other one.

[POSIX](https://pubs.opengroup.org/onlinepubs/009695399/functions/gethostbyaddr.html) says that `gethostbyaddr` only returns a single hostname:
> The gethostbyaddr() function shall return an entry containing addresses of address family type for the host with address addr.

> How it choice the one to report ?

To my understanding there is no specification or standard that say which one it should pick, so it is completely by random, depending on the implementation.

> What about the other ? Why should be ok to ignore it ?

You have a good explanation here why multiple PTR records is a bad idea:
https://serverfault.com/questions/618700/why-multiple-ptr-records-in-dns-is-not-recommended

Basically, by using a different implementation (musl libc) you uncovered a problem or weakness in your network setup. I think you should fix the real problem instead of depending on a "lucky" implementation. You have no guarantees that future versions of glibc, MacOS, windows (or any other implementation) will behave the same.

**Comment by chemelli74** [comment]
No fix on sight...closing.
